Feature/32 feature user persistence store OIDC profile with mariadb

furcev32 opened the pull request:
[Link to Issue.](https://github.com/MoQel/QuaK/issues/32)

furcev32 committed:
First working Implementation of the Google OIDC Login Flow.

furcev32 committed:
Fixed redirecting error after loging out.

furcev32 committed:
Delete unneeded Markdown files.

furcev32 committed:
Merge branch 'feature/35-feature-ui---global-header-navigation-bar' into feature/31-feature-oidc-secure-login-flow-bff

furcev32 committed:
Fix CSRF issues, update API handling for OIDC flow, and configure DB

- Update SecurityConfig to properly handle CSRF tokens and OIDC authentication.
- Update application.properties for security settings.
- Update frontend API utility to include CSRF token in requests.
- Refactor frontend components (CreateDialog, Delete, Directory, etc.) to use updated API methods and handle errors gracefully.
- Update backend tests (FileControllerTest, ProjectControllerTest) to align with security changes.
- Configure docker-compose to automatically update database schema (SPRING_JPA_HIBERNATE_DDL_AUTO).

furcev32 committed:
Fix integration tests by adding CSRF tokens to MockMvc requests

furcev32 committed:
feat: Implement OIDC user sync and update Profile page

- Added User entity and UserRepository
- Implemented OIDC user synchronization service
- Added UserController with /api/me endpoint
- Updated SecurityConfig to enable OIDC login and user sync
- Updated Project and File controllers to enforce ownership
- Rewrote Frontend Profile page to display real user data
- Updated backend tests to use OIDC mocks

furcev32 committed:
chore: Update Docker configuration and documentation

- Updated README with sudo commands for docker compose
- Refactored backend Dockerfile to optionally build frontend
- Updated docker-compose.prod.yaml to bundle frontend in backend
- Added docker-compose.dev.yaml and .dockerignore
- Configured Vite proxy for development

furcev32 committed:
refactor: Update API endpoints to use /api prefix

- Updated FileController and ProjectController base paths
- Updated frontend API calls to match new endpoints

furcev32 committed:
feat: optimize build process and add comprehensive testing infrastructure

## Build Optimizations
- Skip frontend build during backend tests (95% faster: 1m38s → 3-5s)
- Add conditional frontend build via -PbuildFrontend flag
- Add buildProduction task for explicit production builds
- Update Dockerfile to use new build flags

## Testing Infrastructure
- Add Vitest + React Testing Library for frontend testing
- Add root package.json for monorepo test orchestration
- Add Gradle tasks: testAll, testFrontend, testAllSequential
- Add parallel test execution (backend + frontend in ~5-10s)
- Add sample tests and test setup configuration

## Test Endpoints Fix
- Update all test endpoints from /file to /api/file
- Update all test endpoints from /project to /api/project
- All 33 backend tests now passing

## Documentation
- Add comprehensive Testing section to README
- Add Build Commands documentation
- Add examples for writing tests
- Document all test execution options

## Performance Impact
- Backend tests: 1m38s → 3-5s (95% improvement)
- Frontend tests: ~2-3s
- Combined tests (parallel): ~5-10s total
- Production builds: unchanged (~2m with frontend)

Closes #32

furcev32 commented on package-lock.json:
Keeping package-lock.json in your React project makes installs deterministic and avoids “works on my machine” bugs by locking exact dependency versions (including transitive ones). [[1]](https://stackoverflow.com/questions/44297803/what-is-the-role-of-the-package-lock-json)

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/FileController.java:
```diff
@@ -36,24 +43,54 @@
+        if (project == null || project.getOwner() == null || !project.getOwner().getId().equals(user.getId())) {
+            throw new ResponseStatusException(FORBIDDEN, "You do not own this file");
+        }
     }
```
Die aktuelle Implementierung hat eine duplizierte getUser Logik im Controller. Spätestens nach dem Backend Refactoring gehört diese Funktionalität für mich in einen eigenen Service, den du dann injizierst. Dadurch ist die Komponente leichter testbar, der Controller leichter und du schaffst Single Responsibility

furcev32 replied:
getUser logic is now in a UserService class.

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/FileController.java:
```diff
@@ -69,16 +106,25 @@ public FileElement<?> newFile(@RequestBody Map<String, Object> obj, @RequestHead
                 .map(FileElementSaver::getRepository)
                 .flatMap(repo -> repo.findById(fId))
                 .orElseThrow(() -> new ResponseStatusException(BAD_REQUEST, "No matching FileElement found for id"));
+        checkOwnership(element, user);
```
Hier vermischst du Controllerlogik mit Buisnesslogik. Ein Ownership check gehört ebenfalls in einen Service Bsp. SecurityService und sollte nicht Teil von den Controllern sein. Dann kannst du später auch verschiedene Rollen/Gruppen zentral regeln.

furcev32 replied:
Now, This verfication logic is also moved to the UserService class.

furcev32 committed:
#32: Remove redundant dark mode initialization in main.tsx

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/ProjectController.java:
```diff
@@ -28,44 +34,63 @@
+                    .orElseThrow(() -> new ResponseStatusException(FORBIDDEN, "User not found"));
+        }
+        throw new ResponseStatusException(FORBIDDEN, "Not authenticated");
     }
```
Gleiches Spiel wie im FileController, was mir noch aufgefallen ist, dass ich eigene HTTP Error Messages werfen würde 401 für unauthorized und 403 für keinen Zugriff/keine Rechte, das würde im Frontend dann semantisch das klarer halten

furcev32 replied:
Jetzt existiert auch diese Trennung.

furcev32 committed:
#32: Move API usage documentation to top of file

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/FileController.java:
```diff
@@ -36,24 +43,54 @@
+
         final RepoMonad<?> dest = savers.getSaverForElementId(parent)
                                         .flatMap(FileElementSaver::getRepoMonad)
                                         .orElseThrow(() -> new ResponseStatusException(BAD_REQUEST, "No matching parent found"));
```
Diese Methode schreit halt nach Refactoring... Da müssen wir schauen, wie und ob wir das so mergen und wie das mit dem kommenden Backend Refactoring zusammenspielt. Wäre mega schade, wenn man die ganze Funktionalität beim rüberbringen in die neue Architektur quasi neu implementieren müsste.

furcev32 replied:
Jetzt ist die Bussinesslogic in einer seperaten Service Klasse.

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/FileController.java:
```diff
@@ -100,20 +152,24 @@ public void patchFileElement(@PathVariable String fId, @RequestBody Map<String,
         File file = files.findById(fId).orElseThrow(
                 () -> new ResponseStatusException(BAD_REQUEST, "Given file-ID does not resolve to an existing file.")
         );
+        checkOwnership(file, user);
```
In Zukunft sollten wir soetwas mit annotations abdecken z.B. @Authorize als eigene Spring Security-Bean. Das ist zentral wartbar und wiederverwendbar finde ich sogar glaube am aller schönsten. Vielleicht kannst du dir mal anschauen wie das funktioniert . https://docs.spring.io/spring-security/reference/servlet/architecture.html Man könnte das als Spring Filter Chain bauen und da den Bearer Token oder was das für einer ist authentifizieren.

furcev32 replied:
Done!

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/model/FileElement.java:
```diff
@@ -110,4 +110,15 @@ public boolean equals(Object o) {
+        }
+        return null;
+    }
 }
```
Das gehört hier nicht her. Das ist eine Funktion, die lediglich für die Projekt domain gedacht ist und nicht polymorph auf FileElement funktioniert. Ein Tipp am Rande, immer wenn du *instanceof* verwenden musst, sollten deine Alarmglocken klingeln und du solltest doppelt und dreifach überprüfen, ob das wirklich sinnvoll ist.

furcev32 replied:
Wurde gefixt.

NicKIT01 commented on backend/src/main/java/edu/kit/quak/files/ProjectController.java:
```diff
@@ -77,7 +102,8 @@ public Project patchProject(@PathVariable String pId, @RequestBody Project modif
     @Transactional
-    public void deleteProject(@PathVariable String pId) {
+    public void deleteProject(@PathVariable String pId, Authentication authentication) {
+        getProject(pId, authentication); // Check ownership
```
Ich glaube das ist kein guter Stil intern den Controller erneut aufzurufen, vor allem weil wie ich schon angemerkt habe die Überprüfung die du damit machen willst an sich Buisness Logik ist und nicht hierher gehört

furcev32 replied:
Die Buisness Logik ist jetzt in der Separaten ProjectService Klasse

NicKIT01 reviewed (changes requested):
Sorry hab leider nicht geschafft mehr anzuschauen in der Zeit aber da sind auf jeden Fall einige Kommentare an denen du arbeiten kannst und die auf jeden Fall nochmal überarbeitet werden sollten.

Commit:
#32: Update Dockerfile for cross-platform compatibility

Commit:
#32: Refactor backend to hexagonal architecture

## Architecture Changes
- Restructure backend to follow hexagonal/clean architecture principles
- Separate concerns into core, application, and infrastructure layers

## Core Layer (Domain)
- Move all domain models to core/filesystem/model package
- Consolidate User, File, Directory, Project, and related models
- Keep domain logic independent of frameworks

## Application Layer (Use Cases)
- Create service layer in application/filesystem/service
- Add AuthService, UserService, FileService, ProjectService
- Move OidcUserSyncService to application layer
- Implement business logic and orchestration

## Infrastructure Layer (Adapters)
- Move REST controllers to infrastructure/filesystem/in/web/rest
- Relocate FileController, ProjectController, AuthController, UserController
- Controllers now delegate to application services

## Repository Updates
- Update all repository imports to use new core model locations
- Update savers to reference new model packages
- Maintain existing repository functionality

## Test Updates
- Update test imports to match new package structure
- Ensure all 33 tests continue to pass
- Update FileControllerTest and ProjectControllerTest

This refactoring improves maintainability, testability, and follows clean architecture best practices while maintaining all existing functionality.

NicKIT01 commented:
Viel viel besser!! Sehr schön gelöst und auch wirklich gut, dass du die neue Struktur übernimmst. Ich kann morgen nach dem Meeting nochmal ein richtiges Review machen aber auf den ersten Blick sah es gut aus!

Commit:
#32: Expose MariaDB port 3306 in dev docker-compose

simonvonroenn commented:
Diese PR kann geschlossen werden, oder @furcev32?

furcev32 commented:
Moved the PR Progress to the Merge/57 with 32 user persistence #83 branch
